### ISSUE TITLE
QT Core wallet crash when starting with "-txindex"

### ISSUE THREAD

**Issue by user78713** [bitcoin/bitcoin#12406]
Steps to reproduce:

1. Start wallet with "C:\Program Files\Bitcoin\bitcoin-qt.exe" -datadir=d:\BitCoinData -Server "
2. Let it sync 100%
3. Close wallet
4. start again with "-txindex" flag added

The wallet re-sync, but at some point, the wallet just stop syncing, without any error message. When I
close the wallet (press the "X"), error appears "Error: A fatal internal error occurred..."

Attached debug log
[debug.log](https://github.com/bitcoin/bitcoin/files/1713674/debug.log)

Expected behaviour
Wallet should sync, and I should be able to use "getrawtransaction" from debug console

Actual behaviour
Wallet stalls and then just crash when I close it

Screenshots.
![bitcoin_issue2](https://user-images.githubusercontent.com/35986442/36065358-470923ba-0ea2-11e8-91f3-a258068a9426.jpg)
![bitcoin_issue](https://user-images.githubusercontent.com/35986442/36065359-4729c44e-0ea2-11e8-85b8-a8f8551eb1f0.jpg)

**Comment by MarcoFalke** [comment]
See #12401

**Comment by Sjors** [comment]
> at some point, the wallet just stop syncing, without any error message

@MarcoFalke that's a different error than I saw before making #12401

From the end of the log file:

```
2018-02-10 10:07:16 UpdateTip: new best=0000000000000000005e46d334bc42d19a68393a35f323c5661d2e025d942602 height=507987 version=0x20000000 log2_work=88.059978 tx=297969832 date='2018-02-06 17:43:35' progress=0.996700 cache=102.2MiB(770495txo)
2018-02-10 10:07:16 receive version message: /bitnodes.earn.com:0.1/: version 70015, blocks=508490, us=37.142.234.51:8333, peer=84
2018-02-10 10:07:16 net thread exit
2018-02-10 10:07:16 opencon thread exit
2018-02-10 10:07:16 msghand thread exit
2018-02-10 10:07:16 Shutdown: In progress...
2018-02-10 10:07:16 UPNP_DeletePortMapping() returned: 0
2018-02-10 10:07:16 upnp thread interrupt
2018-02-10 10:07:17 IO error: f:\BitCoinData\chainstate\160320.ldb: Data error (cyclic redundancy check).
2018-02-10 10:07:17 *** System error while flushing: Database I/O error
2018-02-10 10:07:35 IO error: f:\BitCoinData\chainstate\160320.ldb: Data error (cyclic redundancy check).
2018-02-10 10:07:35 *** System error while flushing: Database I/O error
2018-02-10 10:07:38 Shutdown: done
```

**Comment by Sjors** [comment]
@user78713 can you attach `BitCoinData\chainstate\160320.ldb` so we can look at the corruption?

**Comment by user78713** [comment]
Renamed to ".txt" so that I can attach it
[160320.ldb.txt](https://github.com/bitcoin/bitcoin/files/1714862/160320.ldb.txt)

**Comment by msafi** [comment]
I ran into a similar error earlier. Bitcoin Qt would crash without logging out anything meaningful. I found out what the problem in my case was by running Bitcoin Qt with the debugger attached. When I did that, the breakpoint was triggered on [this line](https://github.com/bitcoin/bitcoin/blob/f05c1ac444e0c893516535bfdf07c5c8cd9bce16/src/qt/bitcoingui.cpp#L1375). The `message` variable said something about `prune` being turned on while `txindex` is on. I had previously turned on pruning, but I was now trying to run Bitcoin Qt with pruning off by omitting `prune=1` from `bitcoin.conf`. Apparently, omitting `prune` is not enough. I had to explicitly set `prune=0`.

TLDR; setting `prune=0` in `bitcoin.conf` fixed this problem for me.

**Comment by Sjors** [comment]
@msafi were you using a recent version of master or a release? I believe we briefly introduced a bug that could trigger the message assert, but it's been reverted. See #17965 (cc @promag, @hebasto).

**Comment by msafi** [comment]
@Sjors Yes, that may have been the case.

**Comment by MarcoFalke** [comment]
@user78713 
Bitcoin Core makes heavy use of CPU, RAM and disk IO. Hardware defects might only become visible when running Bitcoin Core. You might want to check your hardware for defects.

* memtest86 to check your RAM
* to check the CPU behaviour under load, use linpack or Prime95
* to test your storage device use smartctl or CrystalDiskInfo

Source: https://bitcoin.stackexchange.com/a/12206

**Comment by MarcoFalke** [comment]
Also, Bitcoin Core does not support external drives connected over USB, as they will also cause leveldb corruption.
